### PR TITLE
DENG 946 - Update DAG generation to include ETL checks

### PR DESCRIPTION
**Pull request by alekhyamoz**
For context, please refer to https://mozilla-hub.atlassian.net/browse/DENG-946

This PR includes all the bqetl changes required to include DQ checks in DAG generation.
There is follow-up [PR 1735](https://github.com/mozilla/telemetry-airflow/pull/1735) in `telemetry-airflow` repo that add the `bigquery_dq_check` to gcp utils.

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

### CODE REVIEW

File: bigquery_etl/query_scheduling/generate_airflow_dags.py [hunk 1]
```diff
@@ -52,12 +54,12 @@ def get_dags(project_id, dags_config):
     """Return all configured DAGs including associated tasks."""
     tasks = []
     dag_collection = DagCollection.from_file(dags_config)
-
     for project_dir in project_dirs(project_id):
         # parse metadata.yaml to retrieve scheduling information
         if os.path.isdir(project_dir):
             for root, dirs, files in os.walk(project_dir):
                 try:
+                    checks_task = None
```
Comment (scholtzan) [inline bigquery_etl/query_scheduling/generate_airflow_dags.py:62]:
```suggestion
```
I don't think this is needed. `checks_task` gets set on line 100 and is only used in that scope

File: bigquery_etl/query_scheduling/templates/airflow_dag.j2 [hunk 2]
```diff
@@ -184,12 +229,12 @@ with DAG('{{ name }}', default_args=default_args{%+ if schedule_interval != None
 {% set fivetran_seen = [] -%}
 {% for task in tasks | sort(attribute='task_name') %}
     {% for dependency in (task.upstream_dependencies + task.depends_on) | sort(attribute='task_id') -%}
-    {% if dependency.dag_name == name and dependency not in task.depends_on -%}
+    {% if dependency.dag_name == name and dependency not in task.depends_on-%}
```
Comment (scholtzan) [inline bigquery_etl/query_scheduling/templates/airflow_dag.j2:232]:
```suggestion
    {% if dependency.dag_name == name and dependency not in task.depends_on -%}
```
too keep it more readable

File: dags/bqetl_download_funnel_attribution.py [hunk 3]
```diff
@@ -66,6 +80,10 @@
         pool="DATA_ENG_EXTERNALTASKSENSOR",
     )
 
+    checks__ga_derived__downloads_with_attribution__v2.set_upstream(
+        wait_for_ga_derived__www_site_empty_check__v1
+    )
+
```
Comment (scholtzan) [inline dags/bqetl_download_funnel_attribution.py:86]: mmh, I don't think this is correct? There should only be one upstream dependency which is `ga_derived__downloads_with_attribution__v2`

File: bigquery_etl/query_scheduling/generate_airflow_dags.py [hunk 4]
```diff
@@ -93,6 +94,17 @@ def get_dags(project_id, dags_config):
...
+                        checks_file = os.path.join(root, CHECKS_FILE)
+                        checks_task = copy.deepcopy(
+                            Task.of_dq_check(checks_file, dag_collection=dag_collection)
+                        )
+                        tasks.append(checks_task)
+                        task_ref = TaskRef(
+                            dag_name=task.dag_name,
+                            task_id=task.task_name,
+                        )
+                        checks_task.upstream_dependencies.append(task_ref)
```
Comment (scholtzan) [inline bigquery_etl/query_scheduling/generate_airflow_dags.py:107]: we were using `depends_on` previously but should have used `upstream_dependencies`. `depends_on` is generally using for defining dependencies on Airflow tasks that live in telemetry-airflow (or anything not coming from bigquery-etl)

**Review by scholtzan — APPROVED** [review]

File: bigquery_etl/query_scheduling/templates/airflow_dag.j2 [hunk 5]
```diff
@@ -56,6 +56,51 @@ with DAG('{{ name }}', default_args=default_args{%+ if schedule_interval != None
...
+            {#+ TODO when Airflow is updated to 2.2+ use ds_nodash instead of ds_format -#}
+            source_table={%+ if task.date_partition_offset -%}'{{ task.destination_table }}${% raw %}{{{% endraw %} macros.ds_format(macros.ds_add(ds, {{ task.date_partition_offset }}), "%Y-%m-%d", "%Y%m%d") {% raw %}}}{% endraw %}'
+                            {%+ elif task.destination_table -%}'{{ task.destination_table }}'
+                            {%+ else -%}None
+                            {%+ endif -%},
+            {%+ if task.query_project -%}
+            dataset_id='{{ task.project }}:{{ task.dataset }}',
+            project_id='{{ task.query_project }}',
+            {%+ if not task.query_file_path -%}
+            sql_file_path="sql/{{task.project}}/{{task.dataset}}/{{task.destination_table}}/checks.sql",
```
Comment (gleonard-m) [inline bigquery_etl/query_scheduling/templates/airflow_dag.j2:71]: We can remove `sql_file_path` from the template to have one less parameter in the Task.  `sql_file_path` will default to `{project_id}/{dataset_id}/{source_table}`.  If the parameter is not removed, we can remove the `checks.sql` part.  Having the filename implies that a different filename can be used however the current implementation will only use `checks.sql`.

**Review by gleonard-m — COMMENTED** [review]
I also updated the `dowmloads_with_attribution_v2/checks.sql` file to have a valid threshold.

**Review by scholtzan — APPROVED** [review]

**Review by scholtzan — APPROVED** [review]
🚢

**Pull request merged**